hibernate-core/src/main/java/org/hibernate/query/sqm/consume/spi/BaseSqmToSqlAstConverter.java

```diff
@@ -967,21 +1030,57 @@ public void accept(SqlExpressableType type) {
 		}
 	}
 
+	private void registerParameterResolvedType(SqmParameter expression, AllowableParameterType<?> resolvedType) {
+		if ( sqmParamResolvedTypeMap == null ) {
+			sqmParamResolvedTypeMap = new HashMap<>();
+		}
+
+		sqmParamResolvedTypeMap.put( expression, resolvedType );
```

Comment:
Yes and no...  

What you say is generally true, but is a more general concern with how to deal with multiple types of different specificity.  Not not just parameters.  Think, e.g., binary arithmetic operations such as `a.b + 1` - what is the type of that expression?  So what this really needs in general terms is a way to allow overriding the resolved type with more specific types.

And no, because in your specific example the type of the parameter is not inferred - it's the type of the CAST expression that is inferred.